src/tree/updater_quantile_hist.cc

```diff
@@ -1269,25 +1200,42 @@ GradStats QuantileHistMaker::Builder<GradientSumT>::EnumerateSplit(
 
 template struct QuantileHistMaker::Builder<float>;
 template struct QuantileHistMaker::Builder<double>;
-template void QuantileHistMaker::Builder<float>::PartitionKernel<uint8_t>(
+template void QuantileHistMaker::Builder<float>::Partition<uint8_t, true>(
     const size_t node_in_set, const size_t nid, common::Range1d range,
     const int32_t split_cond, const ColumnMatrix& column_matrix, const RegTree& tree);
-template void QuantileHistMaker::Builder<float>::PartitionKernel<uint16_t>(
+template void QuantileHistMaker::Builder<float>::Partition<uint16_t, true>(
     const size_t node_in_set, const size_t nid, common::Range1d range,
     const int32_t split_cond, const ColumnMatrix& column_matrix, const RegTree& tree);
-template void QuantileHistMaker::Builder<float>::PartitionKernel<uint32_t>(
+template void QuantileHistMaker::Builder<float>::Partition<uint32_t, true>(
     const size_t node_in_set, const size_t nid, common::Range1d range,
     const int32_t split_cond, const ColumnMatrix& column_matrix, const RegTree& tree);
-template void QuantileHistMaker::Builder<double>::PartitionKernel<uint8_t>(
+template void QuantileHistMaker::Builder<double>::Partition<uint8_t, true>(
     const size_t node_in_set, const size_t nid, common::Range1d range,
     const int32_t split_cond, const ColumnMatrix& column_matrix, const RegTree& tree);
-template void QuantileHistMaker::Builder<double>::PartitionKernel<uint16_t>(
+template void QuantileHistMaker::Builder<double>::Partition<uint16_t, true>(
     const size_t node_in_set, const size_t nid, common::Range1d range,
     const int32_t split_cond, const ColumnMatrix& column_matrix, const RegTree& tree);
-template void QuantileHistMaker::Builder<double>::PartitionKernel<uint32_t>(
+template void QuantileHistMaker::Builder<double>::Partition<uint32_t, true>(
```

Comment:
Sorry, maybe I didn't catch the idea. Now it's possible to move implementation of the `Partition` method into header (`PartitionKernel` should be also moved into header file), instantiations can be removed. Is it applicable? Could you add more details about modularization?

Reply:
This is what I had in mind: https://github.com/trivialfis/xgboost/blob/rewrite-cpu-hist/src/tree/hist/row_partitioner.h .  We can do it later since we also need to split up the tests.  Just curious about whether you are interested in this.

Reply:
I can try to move all related to partition methods into `PartitionBuilder` class. Is it what expected?
It seems reasonable to simplify `updater_quantile_hist` files

Reply:
> Is it what expected?

Yup, I will leave it for you to decide whether it should be part of this PR. ;-)

> It seems reasonable to simplify updater_quantile_hist files

I agree that we can cut it down a little bit, thanks for working on this!

Reply:
Partition method was moved to new `partition_builder.h` separate header file. Tests were fixed.
@trivialfis please check it